apps/src/lib/tools/jsdebugger/redux.js

```diff
@@ -0,0 +1,261 @@
+  watchIntervalId: null,
+  commandHistory: null,
+  logOutput: '',
+  jsInterpreterNextStep: null,
+  jsInterpreterPaused: false,
+});
+
+export function getRoot(state) {
+  return state.jsdebugger;
+}
+
+export function getCommandHistory(state) {
+  return getRoot(state).commandHistory;
+}
+
+export function getJSInterpreter(state) {
+  return getRoot(state).jsInterpreter;
+}
+
+function getRunApp(state) {
+  return getRoot(state).runApp;
+}
+
+export function isPaused(state) {
+  return getRoot(state).jsInterpreterPaused;
+}
+
+function getObserver(state) {
+  return getRoot(state).observer;
+}
+
+function getWatchIntervalId(state) {
+  return getRoot(state).watchIntervalId;
+}
+
+export function isAttached(state) {
+  return !!getJSInterpreter(state);
+}
+
+export function canRunNext(state) {
+  return (
+    isPaused(state) &&
+    getRoot(state).jsInterpreterNextStep === JSInterpreter.StepType.RUN
+  );
+}
+
+export function getLogOutput(state) {
+  return getRoot(state).logOutput;
+}
+
+export const selectors = {
+  getRoot,
+  getCommandHistory,
+  getJSInterpreter,
+  isPaused,
+  isAttached,
+  canRunNext,
+  getLogOutput,
+};
+
+// actions
+export function initialize({runApp}) {
+  return {type: INITIALIZE, runApp};
+}
+
+export function appendLog(output) {
+  return {type: APPEND_LOG, output};
+}
+
+export function clearLog() {
+  return {type: CLEAR_LOG};
+}
+
+function mirrorInterpreterState(jsInterpreter) {
+  return {type: MIRROR_INTERPRETER_STATE, jsInterpreter};
+}
+
+export function attach(jsInterpreter) {
+  return (dispatch, getState) => {
+    const observer = new Observer();
+    observer.observe(
+      jsInterpreter.onNextStepChanged,
+      () => dispatch(mirrorInterpreterState(jsInterpreter))
+    );
+    observer.observe(
+      jsInterpreter.onPause,
+      () => dispatch(togglePause())
+    );
+    observer.observe(
+      jsInterpreter.onExecutionWarning,
+      (output) => dispatch(appendLog(output))
+    );
+
+    const watchIntervalId = setInterval(
+      () => {
+        const jsInterpreter = getJSInterpreter(getState());
+        getState().watchedExpressions.forEach(we => {
+          const currentValue = jsInterpreter.evaluateWatchExpression(we.get('expression'));
+          dispatch(updateWatchExpressions(we.get('expression'), currentValue));
+        });
+      },
+      WATCH_TIMER_PERIOD
+    );
+    dispatch(clearLog());
+    dispatch({
+      type: ATTACH,
+      observer,
+      watchIntervalId,
+      jsInterpreter,
+    });
+  };
+}
+
+export function detach() {
+  return (dispatch, getState) => {
+    const observer = getObserver(getState());
+    observer && observer.unobserveAll();
+    clearInterval(getWatchIntervalId(getState()));
+    dispatch({type: DETACH});
+  };
+}
+
+export function togglePause() {
+  return (dispatch, getState) => {
+    const jsInterpreter = getJSInterpreter(getState());
+    if (!jsInterpreter) {
+      return;
+    }
+    jsInterpreter.handlePauseContinue();
```

Comment:
This has made me cringe for a long time but I haven't worked out what to do about it.  Does the interpreter pause the debugger or does the debugger pause the interpreter?  I'm honestly not sure.